Discord: Week 3

LGmatrix13 opened the pull request:
* Made a sub-module for `chatops4s-discord`.
* Create models from the [spec](https://gist.github.com/Krever/eee34f786bfbb9854ae12c17c88889a4).
* Created an implementation of the `trait` `InboundGateway`, `DiscordInbound`
* Created an implementation of the `trait` `OutboundGateway`, `DiscordOutbound`
* Using `sttp.client4` for http request
* Using `tapir` for routing
* Using `cats-effect` for asynchronous execution.   
* Created a test `Server.scala` file that runs an IO-aware program with `http4s` to handle requests via `cats-effect`.

Krever commented on chatops4s-discord/build.sbt:
```diff
@@ -0,0 +1,23 @@
+    libraryDependencies ++= Seq(
+      "com.softwaremill.sttp.tapir" %% "tapir-core"             % "1.11.34",
+      "com.softwaremill.sttp.tapir" %% "tapir-cats"             % "1.11.34",
+      "com.softwaremill.sttp.tapir" %% "tapir-http4s-server"    % "1.11.34",
```
The end goal is for the client to select the backend. Let me elaborate on this during the sync call

Krever commented on chatops4s-discord/src/main/scala/api/DiscordOutbound.scala:
```diff
@@ -0,0 +1,68 @@
+
+  private def baseRequest = basicRequest
+    .header("Authorization", s"Bot $token")
+    .header("User-Agent", s"DiscordBot ($url, $versionNumber)")
```
Is this required by the spec/discord api?

LGmatrix13 replied:
It is. They required you to send along metadata and a `User-Agent` header: https://discord.com/developers/docs/reference#user-agent

Krever commented on chatops4s-discord/src/main/scala/api/DiscordOutbound.scala:
```diff
@@ -0,0 +1,68 @@
+          "components" := message.interactions.map { b =>
+            Json.obj(
+              "type" := 2,
+              "style" := 1,
```
those magic numbers (1,2) are bit worrying. Cany we "document it" by making them variables (can be within `sendToChannel`) with some more meaningful name?

Krever commented on chatops4s-discord/src/main/scala/api/DiscordOutbound.scala:
```diff
@@ -0,0 +1,68 @@
+      .body(json.noSpaces)
+      .response(asJson[Json])
+
+    HttpClientCatsBackend.resource[IO]().use { backend =>
```
In the end we want the backend to be injectable by the user. Let me elaborate on the call.

Krever commented on chatops4s-discord/src/main/scala/api/DiscordOutbound.scala:
```diff
@@ -0,0 +1,68 @@
+      request.send(backend).flatMap { response =>
+        response.body match {
+          case Right(json) =>
+            IO.println("success")
```
This will not run 😅 I will elaborate why on the call. But its a a very typical mistake and very good you made it. Gives me oportunity to talk 😅

Krever replied:
From different angle, we would need a better approach to logging. My usual is scala-logging.

Krever commented on chatops4s-discord/src/main/scala/api/Server.scala:
```diff
@@ -0,0 +1,167 @@
+import sttp.tapir.swagger.bundle.SwaggerInterpreter
+import utilities.EnvLoader
+
+object Server extends IOApp:
```
This one we have to split into three parts:
1. Endpoints definitions 
2. Service with the logic
3. Main that starts the server 

1 and 2 might be together for now (usually they get split if we care about openapi spec, but here we probably dont, as we implement predefined api)

LGmatrix13 commented:
Week 5:

- Added logging with `scala-logging`
- Allowed logging to be turned on via a `verbose` flag
- Replaced magic values with enums
- Removed debugging print statements
- Made server generalized. Removed usage of `http4s` in library and in `build.sbt`
- Made `DiscordOutbound`'s useage of `sttp` generalized by removing usage of a cats backend. 
- Created `chatops4s-discord-example` in project root with some starter files.

Krever commented on chatops4s-discord/src/main/scala/api/DiscordInbound.scala:
```diff
@@ -0,0 +1,29 @@
+import scala.collection.concurrent.TrieMap
+
+class DiscordInbound(verbose: Boolean = false) extends InboundGateway {
+  private final val logger = Logger(getClass.getName)
```
you dont have to create logger by hand. One utility of scala-logging is that you can write `extends StrictLogging` and thats it.

Krever commented on chatops4s-discord/src/main/scala/api/DiscordInbound.scala:
```diff
@@ -1,29 +1,25 @@
 
   override def registerAction(handler: InteractionContext => IO[Unit]): ButtonInteraction = {
-    if (verbose) logger.info("Registered action")
+    logger.whenInfoEnabled {
```
this is already baked in the logger logic. All you need to do is to blindly call `logger.info()`

LGmatrix13 commented:
@Krever still working on wiring up the example as a separate library, but the all the other revisions I have implemented.

Krever commented on chatops4s-discord-example/src/main/scala/services/SendToProductionService.scala:
```diff
@@ -0,0 +1,30 @@
+import com.typesafe.scalalogging.{Logger, StrictLogging}
+
+class SendToProductionService extends StrictLogging {
+  private val discordOutbound = new DiscordOutbound()
```
this you would usually pass through constructor, no need to recreate everywhere its needed.

Krever commented on chatops4s-discord-example/src/main/scala/Main.scala:
```diff
@@ -0,0 +1,77 @@
+    )
+
+
+  private val swaggerRoutes: HttpRoutes[IO] =
```
this can be dropped I think

Krever reviewed (commented):
This time only some minor stuff. We are almost there :D
